src/gridfs/file.rs

```diff
@@ -0,0 +1,641 @@
+    }
+
+    /// Ensures the file mode matches the desired mode.
+    pub fn assert_mode(&self, mode: Mode) -> Result<()> {
+        if self.mode != mode {
+            return match self.mode {
+                Mode::Reading => Err(ArgumentError("File is open for reading.".to_owned())),
+                Mode::Writing => Err(ArgumentError("File is open for writing.".to_owned())),
+                Mode::Closed => Err(ArgumentError("File is closed.".to_owned())),
+            }
+        }
+        Ok(())
+    }
+
+    /// Completes writing or reading and closes the file. This will be called when the
+    /// file is dropped, but errors will be ignored. Therefore, this method should
+    /// be called manually.
+    pub fn close(&mut self) -> Result<()> {
+
+        // Flush chunks
+        if self.mode == Mode::Writing {
+            try!(self.flush());
+        }
+
+        let _ = try!(self.mutex.lock());
+
+        // Complete file write
+        if self.mode  == Mode::Writing {
+            if try!(self.err_description()).is_none() {
+                if self.doc.upload_date.is_none() {
+                    self.doc.upload_date = Some(UTC::now());
+                }
+                self.doc.md5 = self.wsum.result_str();
+                try!(self.gfs.files.insert_one(self.doc.to_bson(), None));
+                //self.gfs.chunks.ensure_index_key("files_id", "n");
+            } else {
+                try!(self.gfs.chunks.delete_many(doc!{ "files_id" => (self.doc.id.clone()) }, None));
+            }
+        }
+
+        // Complete pending chunk pre-load and wipe cache
+        if self.mode == Mode::Reading && self.rcache.is_some() {
+            {
+                let cache = self.rcache.as_ref().unwrap();
+                let _ = try!(cache.lock());
+            }
+            self.rcache = None;
+        }
+
+        self.mode = Mode::Closed;
+
+        let description = try!(self.err_description());
+        if description.is_some() {
+            Err(OperationError(description.unwrap()))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Inserts a file chunk into GridFS.
+    fn insert_chunk(&self, n: i32, buf: &[u8]) -> Result<()> {
+
+        // Start a pending write and copy the buffer and metadata into a bson document
+        self.wpending.fetch_add(1, Ordering::SeqCst);
+        let mut vec_buf = Vec::with_capacity(buf.len());
+        vec_buf.extend(buf.iter().cloned());
+
+        let document = doc! {
+            "_id" => (try!(oid::ObjectId::new())),
+            "files_id" => (self.doc.id.clone()),
+            "n" => n,
+            "data" => (BinarySubtype::Generic, vec_buf)
+        };
+
+        // Insert chunk asynchronously into the database.
+        let arc_gfs = self.gfs.clone();
+        let arc_mutex = self.mutex.clone();
+        let arc_wpending = self.wpending.clone();
+        let cvar = self.condvar.clone();
+        let err = self.err.clone();
+
+        thread::spawn(move || {
```

Comment:
I'm not entirely sure we want to make one part of the driver async and the rest of the driver sync. What's your reasoning for making this async?